[PT Run] Improve the Win32 Program Indexing speed

royvou opened the pull request:
## Summary of the Pull Request

**What is this about:**
Improve the performance of the Win32 scanning.

**What is include in the PR:** 
- Performance Improvements
  - Parallelize collection of all paths with multiple sources
  - Deduplicate all paths which are duplicate   
  - Resolve all indexed program paths in parallel 
    -  When checking Steam/Epic games, do not read whole file if not needed
- Refactored indexing on several points
  - Always LowerCase the fullpath (like done on other places
  - Updated some usage of more modern C# features where i think the code became more readable
- Removed Indexing from constructor (so it won't run if the plugin is disabled)
- Fix Chrome PWA's not having the correct AppType
- Change Security Exceptions to Warnings in the logs 
-  ~~Allow Searching by executable name (e.g. WINWORD.EXE) exluded as this changes behavior~~

**How does someone test / validate:** 

- Run PT Run and check the logs if it's faster for that device
- Disable PT Run within the Settings UI & the 'Microsoft.Plugin.Program.Main - Win32Program index cost ' will not be logged

## Quality Checklist

- [x] **Linked issue:** #7208
- [x] **Communication:** I've discussed this with core contributors in the issue. 
- [x] **Tests:** Added/updated and all pass
- [ ] **Installer:** Added/updated and all pass
- [x] **Localization:** All end user facing strings can be localized
- [ ] **Docs:** Added/ updated
- [ ] **Binaries:** Any new files are added to WXS / YML
   - [x] No new binaries
   - [ ] [YML for signing](https://github.com/microsoft/PowerToys/blob/master/.pipelines/pipeline.user.windows.yml#L68) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/master/installer/PowerToysSetup/Product.wxs) for new binaries

## Contributor License Agreement (CLA)
A CLA must be signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA.


# Additional Information

royvou committed:
Move Indexing Logic to Init

royvou committed:
Refactor Win32 indexing to prevent programs from being indexed twice (from multiple sources) & Optimze Paralell indexing

royvou committed:
Move "IsWebApplication" Check to Indexing

htcfreek commented:
@royvou 
#10981 can conflict with your PR!

enricogior commented:
@royvou 
if you have the time, it would also be great to stop logging as exceptions errors that are not fatal. It's ok to report them as info/warn.

royvou committed:
Implement Grouping by Group

royvou committed:
Log security related exceptions as a warning

royvou commented:
I need to verify the filewatcher logic as the unit tests are failing

royvou committed:
Implement EqualityComparer

royvou committed:
Add setenv

mykhailopylyp commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -403,30 +392,28 @@ private static Win32Program InternetShortcutProgram(string path)
-                            // To catch the exception if the uri cannot be parsed.
-                            // Link to watson crash: https://watsonportal.microsoft.com/Failure?FailureSearchText=5f871ea7-e886-911f-1b31-131f63f6655b
-                            ProgramLogger.Exception($"url could not be parsed", e, MethodBase.GetCurrentMethod().DeclaringType, urlPath);
+                            ProgramLogger.Exception($"url could not be parsed", null, MethodBase.GetCurrentMethod().DeclaringType, urlPath);
```
Why wasn't it changed to Warn? What is the reason for changing the second argument from e to null?

royvou replied:
I've changed The Parse variant to the TryParse variant (so it won't throw an exceptions), this can indeed be changed to a Warn!

mykhailopylyp replied:
It is still Exception

mykhailopylyp commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -885,41 +802,36 @@ private static string GetProgramPathFromRegistrySubKeys(RegistryKey root, string
+                => app1 != null
+                   && app2 != null
+                   && (app1.Name?.ToUpperInvariant(), app1.ExecutableName?.ToUpperInvariant(), app1.FullPath?.ToUpperInvariant())
+                    .Equals((app2.Name?.ToUpperInvariant(), app2.ExecutableName?.ToUpperInvariant(), app2.FullPath?.ToUpperInvariant()));
```
if `app1 == null` and `app2 == null` then the function returns false. Shouldn't it return true?

mykhailopylyp commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -928,42 +840,41 @@ private class RemoveDuplicatesComparer : IEqualityComparer<Win32Program>
-                if (!string.IsNullOrEmpty(app1.Name) && !string.IsNullOrEmpty(app2.Name)
-                    && !string.IsNullOrEmpty(app1.ExecutableName) && !string.IsNullOrEmpty(app2.ExecutableName)
-                    && !string.IsNullOrEmpty(app1.FullPath) && !string.IsNullOrEmpty(app2.FullPath))
+                if (app1 == null || app2 == null)
```
If both apps are null we return false. Shouldn't it be true?

mykhailopylyp commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -928,42 +840,41 @@ private class RemoveDuplicatesComparer : IEqualityComparer<Win32Program>
 
-                return false;
+                return (app1.FullPath.ToUpperInvariant(), app1.HasArguments ? app1.Arguments.ToUpperInvariant() : null)
+                    .Equals((app2.FullPath.ToUpperInvariant(), app2.HasArguments ? app2.Arguments.ToUpperInvariant() : null));
```
Should we handle a case when FullPath or Arguments are null?

royvou replied:
HasArguments does check for null

mykhailopylyp commented:
Program plugin with changes doesn't show any results from the Program plugin.

Master branch:

![image](https://user-images.githubusercontent.com/17161067/119340178-d8031c00-bc9a-11eb-8b81-a72ea4d73da1.png)

Current branch:

![image](https://user-images.githubusercontent.com/17161067/119340004-91adbd00-bc9a-11eb-88a7-0cdbb5eb027f.png)

htcfreek commented:
@enricogior 
Does it make sense to look at #10210 while refactoring plugin code?

mykhailopylyp commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -842,21 +762,18 @@ private static ParallelQuery<Win32Program> AppPathsPrograms(IList<string> suffix
-
-            return filtered;
+            return programs
+                .Where(path => suffixes.Any(suf => path.EndsWith(suf, StringComparison.InvariantCultureIgnoreCase)))
```
New code does not filer by disabledProgramsList. Is it expected?

mykhailopylyp commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -885,41 +802,36 @@ private static string GetProgramPathFromRegistrySubKeys(RegistryKey root, string
         public override bool Equals(object obj)
+            => obj is Win32Program win && Win32ProgramEqualityComparer.Default.Equals(this, win);
+
+        private class Win32ProgramEqualityComparer : IEqualityComparer<Win32Program>
```
Why do we need this comparer?

mykhailopylyp replied:
Before those changes we used only one

mykhailopylyp replied:
By the way, because we use `RemoveDuplicatesComparer` indexing fails and we aren't getting any results on search.

mykhailopylyp commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -976,35 +887,22 @@ public static Win32Program[] All(ProgramPluginSettings settings)
             {
-                var programs = new List<Win32Program>().AsParallel();
-
-                var unregistered = UnregisteredPrograms(settings.ProgramSources, settings.ProgramSuffixes);
```
Can you elaborate on why `UnregisteredPrograms` is deleted?

royvou replied:
This is also functionality which is not yet enabled in PT run (add custom indexing paths)

mykhailopylyp commented:
Measured improvement

**The branch:**
Microsoft.Plugin.Program.Main - Win32Program index cost <796ms>
Microsoft.Plugin.Program.Main - Win32Program index cost <244ms>
Microsoft.Plugin.Program.Main - Win32Program index cost <245ms>

**Master:**
Microsoft.Plugin.Program.Main - Win32Program index cost <280ms>
Microsoft.Plugin.Program.Main - Win32Program index cost <265ms>
Microsoft.Plugin.Program.Main - Win32Program index cost <264ms>

royvou committed:
- Add Static InvalidProgram (To reduce allocations)
- ProgramLogger.Warn allow Exception NULL
- Add "CustomProgramPaths" back
- Moved Enumeration/Distinct
   - ToList on each "ProgramSource"
   - Moved Enumeration/Distinct to All
   - Moved DisabledProgramList check to All
- Changed ToArray to ToList in All to optimize  allocations
- Add InitialSize to HashSet to prevent HashSet resize

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Logger/ProgramLogger.cs:
```diff
@@ -17,6 +17,31 @@ namespace Microsoft.Plugin.Program.Logger
     /// </summary>
     internal static class ProgramLogger
     {
+        /// <summary>
```
Added Warn Method which could be used for non-fatal index related issues (e.g. Not found / not accessible / not parse able)

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Main.cs:
```diff
@@ -118,6 +104,20 @@ public void Init(PluginInitContext context)
 
             UpdateUWPIconPath(_context.API.GetCurrentTheme());
+
+            var a = Task.Run(() =>
```
Moved from Constructor to Initialize method so it won't run if you disable the program plugin

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -28,6 +27,8 @@ namespace Microsoft.Plugin.Program.Programs
     [Serializable]
     public class Win32Program : IProgram
     {
+        public static readonly Win32Program InvalidProgram = new Win32Program { Valid = false, Enabled = false };
```
Single Instance Method (which can be reused to reduce allocations)

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -53,7 +54,7 @@ public class Win32Program : IProgram
         public bool Enabled { get; set; }
 
-        public bool HasArguments { get; set; }
+        public bool HasArguments => !string.IsNullOrEmpty(Arguments);
```
This parameter can easily be forgotten, this will calculate the HasArgument at runtime

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -104,10 +105,12 @@ public bool IsWebApplication()
             // To Filter PWAs when the user searches for the main application
             // All Chromium based applications contain the --app-id argument
-            // Reference : https://codereview.chromium.org/399045/show
+            // Reference : https://codereview.chromium.org/399045
```
The old link was throwing a 404 :(

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -119,9 +122,6 @@ public bool FilterWebApplication(string query)
                 return false;
             }
 
-            // Set the subtitle to 'Web Application'
```
Moved this check inside the LnkProgram

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -403,65 +395,62 @@ private static Win32Program InternetShortcutProgram(string path)
                         urlPath = line.Substring(urlPrefix.Length);
 
-                        try
+                        if (!Uri.TryCreate(urlPath, UriKind.RelativeOrAbsolute, out Uri _))
```
Use "Safe" TryParse (which does not throw an exception, but we log the warning anyway for debugging purpose

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -364,35 +355,36 @@ private static Win32Program CreateWin32Program(string path)
             }
         }
 
+        private static readonly Regex InternetShortcutURLPrefixes = new Regex(@"^steam:\/\/(rungameid|run)\/|^com\.epicgames\.launcher:\/\/apps\/", RegexOptions.Compiled);
```
Regex can be cached/reused

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -403,65 +395,62 @@ private static Win32Program InternetShortcutProgram(string path)
                         ExecutableName = Path.GetFileName(path),
                         IcoPath = iconPath,
-                        FullPath = urlPath,
+                        FullPath = urlPath.ToLowerInvariant(),
```
All Other Win32Programs use LowerInvariant

mykhailopylyp replied:
Can we change it ToUpperInvariant as it is recommended?

royvou replied:
Since its shown to the user (and all other types also use the lowercase variant) I don't think we should!

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -472,33 +461,35 @@ private static Win32Program LnkProgram(string path)
-                        }
-                        else
+                    // Allow searching by executable name (E.g. WINWORD.exe for word)
+                    program.ExecutableName = Path.GetFileName(target);
```
E.g. Word has the .executable name WINWORD, allow search for this.

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -472,33 +461,35 @@ private static Win32Program LnkProgram(string path)
-                        else
+                    // Allow searching by executable name (E.g. WINWORD.exe for word)
+                    program.ExecutableName = Path.GetFileName(target);
+                    program.Arguments = ShellLinkHelper.Arguments;
```
Set Arguments, this is used for the IsWebApplication (Chrome PWA) check

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -472,33 +461,35 @@ private static Win32Program LnkProgram(string path)
+
+                    // A .lnk could be a (Chrome) PWA, set correct AppType
+                    program.AppType = program.IsWebApplication()
+                        ? ApplicationType.WebApplication
```
Old file 123

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -597,37 +584,35 @@ public static Win32Program GetAppFromPath(string path)
-            {
-                app = InternetShortcutProgram(path);
+            Win32Program app;
+            switch (GetAppTypeFromPath(path))
```
Add  all Cases to make it obvious how special AppTypes are resolved.

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -885,89 +801,78 @@ private static string GetProgramPathFromRegistrySubKeys(RegistryKey root, string
                 }
 
-                return false;
+                return app1 != null
```
Removed ExecutableName from Equality, as this is always the same as the last part of the FullPath

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -885,89 +801,78 @@ private static string GetProgramPathFromRegistrySubKeys(RegistryKey root, string
-                int namePrime = 13;
-                int executablePrime = 17;
-                int fullPathPrime = 31;
+                => (obj.Name?.ToUpperInvariant(), obj.FullPath?.ToUpperInvariant()).GetHashCode();
```
C# Tuples to let compiler generate hashCode

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -976,36 +881,45 @@ public static Win32Program[] All(ProgramPluginSettings settings)
+                    .AsParallel()
+                    .SelectMany(source => source.IsEnabled ? source.GetPaths() : Enumerable.Empty<string>())
+                    .Where(programPath => disabledProgramsList.All(x => x.UniqueIdentifier != programPath))
+                    .Where(path => !ExecutableApplicationExtensions.Contains(Extension(path))));
```
Rule 965 in previous version

royvou commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -885,89 +801,78 @@ private static string GetProgramPathFromRegistrySubKeys(RegistryKey root, string
 
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1031:Do not catch general exception types", Justification = "Keeping the process alive but logging the exception")]
-        public static Win32Program[] All(ProgramPluginSettings settings)
+        public static IList<Win32Program> All(ProgramPluginSettings settings)
```
This is build the following way
- Get All Paths (in parallel) inside a hashSet so we only resolve each path once.
- For all Paths, remove the ones which are inside the disabledProgramsList
- Remove Paths we know we don't need to resolve (#965 in old code)
- For each Path (in parallel), resolve based on extension.

mykhailopylyp commented on src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs:
```diff
@@ -818,43 +817,31 @@ public override int GetHashCode()
+                return app1 != null
+                       && app2 != null
+                       && (app1.Name?.ToUpperInvariant(), app1.FullPath?.ToUpperInvariant())
+                       .Equals((app2.Name?.ToUpperInvariant(), app2.FullPath?.ToUpperInvariant()));
```
Can you elaborate on why ExecutableName is removed here?

royvou replied:
As executablename is always set to the last part of the FullPath!

mykhailopylyp replied:
It's not always true. For example, `LnkProgram`:

![image](https://user-images.githubusercontent.com/17161067/119821489-4d6d2780-befb-11eb-8f03-7b65c5a47de8.png)

Also, I found one pair of programs on my machine that have the same FullPath and different ExecutalbeName. I would suggest to return `ExecutableName`.

royvou replied:
Should we not expand the executable name (so you can also use if for the direct exe match), I don't think anyone wants to search for that name?

mykhailopylyp replied:
Let's concentrate on speeding it up without changing behavior where it is possible.

mykhailopylyp commented:
@royvou
On the master, I have 972 indexed programs and on the current branch 965. Any idea why the number differs? I haven't figured it out.

royvou committed:
Correctly resolve full program.ExecutableName

royvou commented:
> 
> 
> @royvou
> On the master, I have 972 indexed programs and on the current branch 965. Any idea why the number differs? I haven't figured it out.

I just validated and i noticed the difference locally is two instances of

 c:\windows\system32\dfrgui.exe
- One resolved as RunCommand
- One resolved as .Lnk (from the startmenu)

The change in the LnkProgram to override the ExeProgram (e.g. to allow to search for WINWORD.exe instead of word.lnk) (#10210) and return back to the 'old behaviour' (Also means program.ExecutableName = Path.GetFileName(program.FullPath); is always true and removes the need to compute the hash of each "ExecuteableName")

royvou committed:
Change type to List

htcfreek commented:
@royvou 
Which file types are supported?

royvou commented:
> 
> 
> @royvou
> Which file types are supported?

The same as before, the only difference is that the combination name <> FullPath should be unique (e.g. a lnk/exe with the same name will be deduplicated.

htcfreek commented:
> > 
> > 
> > @royvou
> > Which file types are supported?
> 
> The same as before, the only difference is that the combination name <> FullPath should be unique (e.g. a lnk/exe with the same name will be deduplicated.

1. What happened if parameters in two lnks are different. (e.g. `program.exe -admintool` & `program.exe -backuptool`)
2. In the code (productive master branch) is a list of file types that includes `.msc`. Should RSAT tools and other mmc tools not be shown then?

royvou commented:
> 
> 
> > > @royvou
> > > Which file types are supported?
> > 
> > 
> > The same as before, the only difference is that the combination name <> FullPath should be unique (e.g. a lnk/exe with the same name will be deduplicated.
> 
>     1. What happened if parameters in two lnks are different. (e.g. `program.exe -admintool` & `program.exe -backuptool`)
> 
>     2. In the code (productive master branch) is a list of file types that includes `.msc`. Should RSAT tools and other mmc tools not be shown then?

1. The change regarding the Arguments Property, it's only used to determine if its a Chrome/Chromium (Not edge) based PWA and show the correct subtitle. Otherwise the logic remains the same (e.g one or the other one will be chosen / the other one deduplicated like now, since some programs have like a -startmenu or -desktop parameter to show where it is launched from)
2. You mean the 'ExecutableApplicationExtensions' variable?

htcfreek commented:
> 2. You mean the 'ExecutableApplicationExtensions' variable?

@royvou
Yes. But the reason why I asked is answered by my self. Please see https://github.com/microsoft/PowerToys/issues/3491#issuecomment-852303293.

royvou commented:
Ah sorry @htcfreek ! (Yes i can search on those names in this PR)

![image](https://user-images.githubusercontent.com/16852398/120367543-4b4b0480-c311-11eb-8c80-a91ef45eac7d.png)
![image](https://user-images.githubusercontent.com/16852398/120367594-59008a00-c311-11eb-9a14-1d267cbabedc.png)

htcfreek commented:
> Ah sorry @htcfreek ! (Yes i can search on those names in this PR)
> 
> ![image](https://user-images.githubusercontent.com/16852398/120367543-4b4b0480-c311-11eb-8c80-a91ef45eac7d.png)
> ![image](https://user-images.githubusercontent.com/16852398/120367594-59008a00-c311-11eb-9a14-1d267cbabedc.png)

Which local do you have? Other than englisch? Maybe you see the reason why some tools not searchable by the translated name while working on this PR.

royvou commented:
> 
> 
> > Ah sorry @htcfreek ! (Yes i can search on those names in this PR)
> > ![image](https://user-images.githubusercontent.com/16852398/120367543-4b4b0480-c311-11eb-8c80-a91ef45eac7d.png)
> > ![image](https://user-images.githubusercontent.com/16852398/120367594-59008a00-c311-11eb-9a14-1d267cbabedc.png)
> 
> @royvou
> Which local do you have? Other than englisch? Maybe you see the reason why some tools not searchable by the translated name while working on this PR.

EN-UK, but this seems not to have changed over the current PT Run release, so let's continue that in the other issue :)

htcfreek commented:
> > 
> > 
> > > Ah sorry @htcfreek ! (Yes i can search on those names in this PR)
> > > ![image](https://user-images.githubusercontent.com/16852398/120367543-4b4b0480-c311-11eb-8c80-a91ef45eac7d.png)
> > > ![image](https://user-images.githubusercontent.com/16852398/120367594-59008a00-c311-11eb-9a14-1d267cbabedc.png)
> > 
> > @royvou
> > Which local do you have? Other than englisch? Maybe you see the reason why some tools not searchable by the translated name while working on this PR.
> 
> EN-UK, but this seems not to have changed over the current PT Run release, so let's continue that in the other issue :)

Issue exists since while. :)

crutkas commented:
@royvou / @mykhailopylyp what is missing to close this out

mykhailopylyp commented:
@crutkas @royvou
I'm good with it as long as we keep win32 program hash function as it was before(that is with executable name) and a number of indexed programs don't differ after this change.

royvou committed:
Add .ExecutableName?.ToUpperInvariant() to Win32ProgramEqualityComparer

royvou committed:
Rollback change with .lnk/exe (resolve winword)

royvou commented:
@mykhailopylyp / @crutkas I'll revert the change that .lnks will always have the .exe name set to the name of the exe (e.g. WINWORD for word) and we should create a seperate issue for that how we want that to behave.
Now it will be the same as on master (e.g. no results)

Also added back the previous comparer.

royvou committed:
Add translations

mykhailopylyp reviewed (approved):
LGTM, Good job! Tested, works as expected
